lib/msf/core/windows_version.rb

```diff
@@ -140,6 +220,29 @@ def xp_or_2003?
       build_number.between?(XP_SP0, Server2003_SP2)
     end
 
+    # Get the string representation of the OS, given a major, minor and build number
+    # (as reported by an NTLM handshake).
+    # The NTLM structure makes no guarantee that it is actually Windows, so if we
+    # don't find a precise match, return nil
+    # @param major [Integer] The major build number reported in the NTLM handshake
+    # @param minor [Integer] The minor build number reported in the NTLM handshake
+    # @param build [Integer] The build build number reported in the NTLM handshake
+    # @return [String] The possible matching OS versions, or nil if no identical match can be found
+    def self.from_ntlm_os_version(major, minor, build)
+      workstation_string = self.version_string(major, minor, build, WorkstationSpecificVersions, WorkstationNameMapping)
+      server_string = self.version_string(major, minor, build, ServerSpecificVersions, ServerNameMapping)
+
+      version_strings = []
+      version_strings.append(workstation_string) unless workstation_string.nil?
+      version_strings.append(server_string) unless server_string.nil?
+
```

Comment:
Do you know if there's a way to detect server vs workstation for at least some of these versions that we could provide as a hint here?

I'm wondering if there's a pipe or share that's present we could use, even if it's only for versions 10+ which is where the numbers start to get shared IIRC.

Reply:
> We don't need to force SMB1 to get the OS information any more.

At the moment, it still does use SMB for that because you'll get the exact OS string from it. So if there are weird versions that we don't have a version number for, it might possibly give better info.

Reply:
> I'm wondering if there's a pipe or share that's present we could use, even if it's only for versions 10+ which is where the numbers start to get shared IIRC.

Not sure - I'll have a look into it.

The version numbers were shared all the way from Vista/Server 2008 up to Win10_1809/Server 2019. After that they _seem_ to be keeping them separate.